gateway/src/lib.rs

```diff
@@ -787,6 +797,101 @@ impl Actor {
         Ok(())
     }
 
+    /// Submit a new cron checkpoint
+    ///
+    /// It only accepts submission at multiples of `cron_period` since `genesis_epoch`, which are
+    /// set during construction. Each checkpoint will have its number of submissions tracked. The
+    /// same address cannot submit twice. Once the number of submissions is more than or equal to 2/3
+    /// of the total number of validators, the messages will be applied.
+    ///
+    /// Each cron checkpoint will be checked against each other using blake hashing.
+    fn submit_cron(rt: &mut impl Runtime, params: CronCheckpoint) -> Result<RawBytes, ActorError> {
+        // submit cron can only be performed by signable addresses
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+
+        let msgs = rt.transaction(|st: &mut State, rt| {
+            // first we check the epoch is the correct one
+            let genesis_epoch = st.genesis_epoch;
+
+            // we process only it's multiple of cron_period since genesis_epoch
+            if (params.epoch - genesis_epoch) % st.cron_period != 0 {
+                return Err(actor_error!(illegal_argument, "epoch not allowed"));
+            }
+
+            let store = rt.store();
+            let submitter = rt.message().caller();
+
+            st.cron_submissions
+                .modify(store, |hamt| {
+                    let epoch_key = BytesKey::from(params.epoch.to_be_bytes().as_slice());
+                    let mut submission = match hamt.get(&epoch_key)? {
+                        Some(s) => s.clone(),
+                        None => CronSubmission::new(store)?,
+                    };
+
+                    if submission.is_executed() {
```

Comment:
I don´t think we need this check, right? You just check if the epoch received is below the last one executed.

Reply:
removed